Macos builder: Performance , bugfixes.

leamas opened the pull request:
Basically fixes for problems described in #2049 

XCode is updated to 11.0, modern enough to avoid the python2 dep (but still not that modern that it exposes CPack bugs when using for example 11.3).

The dependencies are cached, and the build now completes in about 15 minutes instead of 25. Still not "fast", but better. The cache ins invalidated when ci/generic-build-macos.sh is changed.

Libraries used are properly linked to the executable so that CPack can pick them up.

Remove various hacks for library inclusion and chmod.

EDIT: Drop the travis macos builder, use circleci instead. Circleci builds in about half the time, 15 minutes vs 30. 

Closes: #2049

leamas commented:
Added new commits exploring and finally dropping the travis macos build. Having the trials with travis caching in the git history might possible become useful.

rgleason commented:
Alec , why not use config/circleci.yml too, rather than travis?

leamas commented:
This rather long series actually ends up in using circleci instead of travis.

This is just performance and bugfixes besides one crucial point: the update to XCode 11.0. If I understand this correctly, this should be fine if done first in OpenCPN and after that in the plugins since the Xcode level should be backwards compatible.

However, if possible we would need some input from @nohal on this, MacOS is more or less dark arts for me.

To repeat: the reason to update is that current level has dependencies on python2 which is EOL 2021-01-01. The  reason to stay at 11.0 instead of for example 11.3 is that later versions exposes CPack bugs. 

Pavel: thoughts ?

leamas commented:
@rgleason: If you are interested in updating plugins along these line you could look into rasbats/shipdriver_pi instead. This is plugin code and should be easier to use in contexts like testplugin.

rgleason commented:
OK, what about this?   Alec , why not use config/circleci.yml too, rather than travis?

Opensource docker images are retained 6 months  
"Free accounts may retain inactive images for up to 6 months"
https://github.com/docker/roadmap/issues/152

rgleason commented:
OK, I've signed up there. Do I need to download the Docker Desktop for Windows so that I can use the Command Line Interface?
I am in the get started sequence  which looks a bit like github and circleci  (download, clone, build, run, ship).

rgleason commented:
They allow organizations  https://hub.docker.com/orgs   
This has been very useful on Cloudsmith, in fact it is probably the best way to do it due to recent changes at CS.

rgleason commented:
Am I going to be using a pre created circleci container image for MacOS?
Or am I going to have to create my own MacOS image and use Docker CLI to upload it to my repository so that it can be used?

bdbcat commented:
Rick...
There should be no need for you to build/maintain a private MacOS docker image.  And you should not need the docker image locally unless you intend to build and test for MacOS locally.  Do you so intend?

leamas commented:
> OK, what about this? Alec , why not use config/circleci.yml too, rather than travis?

I do, check the code. 

As I said earlier, these commits are not really usable in a plugin context. Look at ShipDriver instead to find plugin code.

rgleason commented:
In answer to Bdbcat,  no, if I don't have to I won't.  If I can have access to the right docker image to build MacOS XCode 11.0 that would be very helpful.

bdbcat commented:
Rick...
As example, the Android plugin builds, and the rPI Raspian builds, are using unmodified public docker images today.  All of this becomes transparent to the plugin devs once the setup is created and validated.
So let's let Alec finish his analysis, then we can implement on the frontend workflows quickly.

rgleason commented:
Alec,  not intending to push or anything..,   I've looked at shipdriver and it seems to be using old MacOS code

-DCMAKE_OSX_DEPLOYMENT_TARGET=10.9 \

and I don't see a docker path and container or username and password.

So I'll wait till later.  Thanks for the heads up..

rgleason commented:
I take that back, wrong repos!    Thanks.
https://github.com/Rasbats/shipdriver_pi/blob/master/.circleci/config.yml
```
 build-macos:
    macos:
      xcode: "11.0.0"
    environment:
      - OCPN_TARGET: macos
    steps:
      - checkout
      - restore_cache:
          keys:
            - macos-cache-v1-{{checksum "ci/circleci-build-macos.sh"}}
      - run: ci/circleci-build-macos.sh
      - save_cache:
          key: macos-cache-v1-{{checksum "ci/circleci-build-macos.sh"}}
```

rgleason commented:
Ok, I'll take a break on the plugins until we get things sorted out more.
I am trying to keep a comprehensive list of the necessary changes to frontend 
https://github.com/OpenCPN/plugins/issues/262

leamas commented:
Rebased to current master.

Added commits for docker authentication.  Uses builder environment variables DOCKER_USER and DOCKKER_PW ; use credentials from account registered on docker.io.

The security here is minimal. I would not use an account on docker.io where I had some important work (images). Better then to register a separate account for the purposes here.

Missing a natural point to document things like these. The manual is not the answer for these kind of docs, they should be part of the git tree.

leamas commented:
This PR is now updated with input from #2121. It 
  - moves the macos builds to circleci,
  - updates the Xcode to 11.3 (necessary, older ersions are not maintained)
  - Updates download links as of #2121 
  - Caches builds deps to speed up the builds
  - Some general clean up, used deps are not linked etc.

The PR fulfills it's basic task to build MacOS. However, here are two regressions, possibly related.
  - The create-pkg target hangs (timeout) when trying to create a .dpg package. I have conditionalized this build on a environment variable, effectively disabling this build for now. Do we need it?
  - The applescript run at the end of create-dmg also hangs, eventually leading to a timeout. Once upon a time, this happened on Travis but not circleci. Now, things are the other way around. I have no idea how to debug this. 

Unfortunately, this is a lot of commits which have evolved over time. @bdbcat: If/when this is merged you might want to use the squash option when merging.

@nohal: @bdbcat: thoughts?

Closes: #2121

leamas commented:
Basic testing of generated dmg is OK: it can be installed and started, nothing wrong at a glance.

leamas commented:
The travis build  will not complete, I'm out of credits. CircleCI build of same commit df78e64d1 : https://app.circleci.com/pipelines/github/leamas/OpenCPN/1184/workflows/32a8de50-1cb6-4a41-8c70-68e9ea8946da/jobs/2268

bdbcat commented:
For MacOS, "create-pkg" is the primary build target.  "create-dmg" is an auxiliary target, kept around for historical reasons.
Attempting solo build of "create-pkg" now, will advise results.

bdbcat commented:
Tail of the logfile:
```
[100%] create-pkg: Done.
[100%] Built target create-pkg
/usr/local/Cellar/cmake/3.19.1/bin/cmake -E cmake_progress_start /Users/distiller/project/build/CMakeFiles 0
+ pip3 install -q cloudsmith-cli

CircleCI received exit code 0

As can be seen, the "create-pkg" target runs fine.  But the install of cloudsmith-cli seems to fail/abort, and nothing is copied.  No idea why.
```

leamas commented:
@bdbcat : Thanks. Seems like an ordering problem: make create-dmg has side-effects and must be run after make pkg. PR under way (just a minimal one, almost everything is already merged).

leamas commented:
links unavailable:

```
 wget  https://download.opencpn.org/s/rwoCNGzx6G34tbC/download -O /tmp/wx312B_opencpn50_macos109.tar.xz
--2020-12-09 09:51:59--  https://download.opencpn.org/s/rwoCNGzx6G34tbC/download
Resolving download.opencpn.org (download.opencpn.org)... 168.119.3.125, 2a01:4f8:242:5dc4::2
Connecting to download.opencpn.org (download.opencpn.org)|168.119.3.125|:443... connected.
HTTP request sent, awaiting response... 503 Service Temporarily Unavailable
2020-12-09 09:51:59 ERROR 503: Service Temporarily Unavailable.
```

Why don't we just set up a cloudsmith repo for our build dependencies? CS seems  pretty stable and has good administrative interfaces, both GUI and CLI. Or some other service like bintray?

leamas commented:
> As can be seen, the "create-pkg" target runs fine. But the install of cloudsmith-cli seems to fail/abort, and nothing is copied. No idea why

It does? Doesn't it  just run fine, no error messages and suppressed diagnostics due to the -q option?

rgleason commented:
Why don't we just set up a cloudsmith repo for our build dependencies? CS seems pretty stable and has good administrative interfaces, both GUI and CLI. Or some other service like bintray?

OK, I will do that  Three files are located at
https://cloudsmith.io/~opencpn/repos/x-buildfiles/packages/  
https://cloudsmith.io/~opencpn/repos/x-buildfiles/packages/
CloudSmith OpenCPN Organization Plugin Team has access to these files and should perhaps be able to update or add files.
If you want to be added let us know.
The repository folder has an "x" so that it lands at the end and does not mess up the 3 folder arrangement of plugins.

leamas commented:
@rgleason: The links are currently handled by Dave and Pavel. If something should be done, it might make sense that they somehow give a hint on how they look at this before rushing away. If they don't consider this a great idea, you have created some sort of fork which will be hard to maintain. 

If they eventually *do* consider this being a good idea they might have input on how it should be implemented.

Bottom line: please hold you horses ;)

rgleason commented:
OK, It is done. It isn't hard to undo.

bdbcat commented:
re:
"It does? Doesn't it just run fine, no error messages and suppressed diagnostics due to the -q option?"
Maybe.  However, no resulting .pkg file ever gets copied to cloudsmith.  So something is not right...

leamas commented:
@bdbcat: could you link to the complete build log?

leamas commented:
@bdbcat: never mind about the build log.

It turns out that the pkg problem was a caching one, create-pkg failed with deps from cache. Fixed in last, rebased commit. From my side, this is now ready to be merged.

Circleci build logs for same commit ff8d040492 at https://app.circleci.com/pipelines/github/leamas/OpenCPN/1199/workflows/9581fb9a-2415-41a7-a186-4c9b42cf5745/jobs/2329 (not cached) and https://app.circleci.com/pipelines/github/leamas/OpenCPN/1199/workflows/686cad48-1a8f-4f7a-be55-8fa4408189e4/jobs/2330 (cached).

In both cases both the .pkg and the .dmg file are uploaded.

bdbcat commented:
The log in master now:
https://app.circleci.com/pipelines/github/OpenCPN/OpenCPN/82/workflows/71bd2b2a-7bfd-4e28-b32d-35276302770c/jobs/206

OK, My bad.  The upload is failing due to missing key.
"Cannot deploy to cloudsmith: missing $CLOUDSMITH_API_KEY"
Fixed, for all CircleCI jobs.
So, I think we can say that the .pkg build is fine.
Now on to the .dmg?

bdbcat commented:
OK. Waiting on travis CI build to complete.  Will merge after....

leamas commented:
The .dmg is already in place...

leamas commented:
Unfortunately, the Travis build will never complete; it runs on my behalf, and I'm out of credits. Perhaps you could merge to a temporary branch to make it build using your credits?

bdbcat commented:
I'll just merge, and we shall see.
Travis build queue seems very long, build takes a while to start.

bdbcat commented:
Merged.
CircleCI build of MacOS completes and uploads ,dmg and .pkg successfully. NOT smoke-tested on real hardware...
Travis builds very slow...
So, call this done?

leamas commented:
> So, call this done?

Yes. I have tested the dmg packages to the point that they can be installed and that OpenCPN and "looks sound". Let's consider it closed.

I'm currently rebuilding #2122 against master; that should be the last change in the ci mess for  now. Will  send a heads-up whne it completes.

leamas commented:
> So, call this done?

Yes. I have tested the dmg packages to the point that they can be installed and that OpenCPN and "looks sound". Let's consider it closed & done.
